File: Python/MinimumCostForTickets.py
import bisect

class Solution:

    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dp = [0] * len(days)
        dp[0] = min(costs)
        for i in range(1, len(days)):
            c1 = dp[i-1] + costs[0]

            d = days[i] - 7
            ret = bisect.bisect(days, d)
            if ret > 0 and days[ret-1] == d:
                c2 = dp[ret-1] + costs[1]
            else:
                if ret == 0:
                    c2 = costs[1]
                else:
                    c2 = dp[ret-1] + costs[1]

            d = days[i] - 30
            ret = bisect.bisect(days, d)
            if ret > 0 and days[ret-1] == d:
                c3 = dp[ret-1] + costs[2]
            else:
                if ret == 0:
                    c3 = costs[2]
                else:
                    c3 = dp[ret-1] + costs[2]
            
            dp[i] = min(c1, c2, c3)
        return dp[-1]

File: Java/src/MinimumCostForTickets.java
import java.util.Arrays;

public class MinimumCostForTickets {
	
	public int mincostTickets(int[] days, int[] costs) {
        int[] dp = new int[days.length];
        
        dp[0] = Math.min(costs[0], Math.min(costs[1], costs[2]));
        for(int i = 1; i < days.length; i++) {
        	int c1 = dp[i-1]+costs[0];
        	
        	int c2 = 0;
        	int d = days[i]-7;
        	int s = Arrays.binarySearch(days, d);
        	if(s >= 0) {
        		c2 = dp[s] + costs[1];
        	}else {
        		int ip = -s-1;
        		if(ip == 0) c2 = costs[1];
        		else c2 = dp[ip-1]+costs[1];
        	}
        	
        	int c3 = 0;
        	d = days[i]-30;
        	s = Arrays.binarySearch(days, d);
        	if(s >= 0) {
        		c3 = dp[s] + costs[2];
        	}else {
        		int ip = -s-1;
        		if(ip == 0) c3 = costs[2];
        		else c3 = dp[ip-1]+costs[2];
        	}
        	
        	dp[i] = Math.min(c1, Math.min(c2, c3));
        }
        
        return dp[days.length-1];
    }
	
	public static void main(String[] args) {
		MinimumCostForTickets test = new MinimumCostForTickets();
		int[] days = {1,2,3,4,5,6,7,8,9,10,30,31};
		int[] costs = {2,7,15};
		System.out.println(test.mincostTickets(days, costs));
	}
	
}
